OpenRA.Mods.Common/Widgets/ObserverArmyIconsWidget .cs

```diff
@@ -0,0 +1,253 @@
+		{
+			GetPlayer = other.GetPlayer;
+			world = other.world;
+			worldRenderer = other.worldRenderer;
+
+			IconWidth = other.IconWidth;
+			IconHeight = other.IconHeight;
+			IconSpacing = other.IconSpacing;
+			iconSize = new float2(IconWidth, IconHeight);
+
+			MinWidth = other.MinWidth;
+
+			TooltipUnit = other.TooltipUnit;
+			GetTooltipUnit = () => TooltipUnit;
+
+			TooltipTemplate = other.TooltipTemplate;
+			TooltipContainer = other.TooltipContainer;
+
+			tooltipContainer = Exts.Lazy(() =>
+				Ui.Root.Get<TooltipContainerWidget>(TooltipContainer));
+
+			Init();
+
+			actorDataCache = new Cache<string, ActorData>(LoadActorDataCache);
+		}
+
+		void Init()
+		{
+			var buildable = world.Map.Rules.Actors.Values
+				.Where(x => x.HasTraitInfo<BuildableInfo>())
+				.SelectMany(x => x.TraitInfo<BuildableInfo>().Queue, (ai, q) => new Tuple<string, ActorInfo, BuildableInfo>(q, ai, ai.TraitInfo<BuildableInfo>()));
+
+			var pqueue = world.Map.Rules.Actors.Values
+				.SelectMany(a => a.TraitInfos<ProductionQueueInfo>());
+
+			unitLookup = buildable.Join(pqueue, tuple => tuple.Item1, pq => pq.Type, (tuple, info) => new { tuple.Item2.Name, Ai = tuple.Item2, tuple.Item3.BuildPaletteOrder, info.DisplayOrder })
+				.Distinct()
+				.ToDictionary(t => t.Name, o => new Tuple<ActorInfo, int, int>(o.Ai, o.DisplayOrder, o.BuildPaletteOrder));
+		}
+
+		public override void Draw()
+		{
+			armyUnits.Clear();
+			armyIconsBounds.Clear();
+
+			var player = GetPlayer();
+			if (player == null)
+				return;
+
+			var playerStatistics = stats.Update(player);
+
+			var items = playerStatistics.Units
+				.Where(i => i.Value.Count > 0)
+				.OrderBy(i => unitLookup[i.Key].Item2)
+				.ThenBy(i => unitLookup[i.Key].Item3)
+				.ToList();
+
+			Bounds.Width = items.Count * (IconWidth + IconSpacing);
+
+			var queueCol = 0;
+			foreach (var unit in items)
+			{
+				var name = unit.Key;
+				var actorInfo = unitLookup[name].Item1;
+
+				var actorData = actorDataCache[name];
+
+				var rsi = actorData.Rsi;
+				var icon = new Animation(world, rsi.GetImage(actorInfo, world.Map.Rules.Sequences, player.Faction.Name));
+
+				icon.Play(actorData.BuildableInfo.Icon);
+				var topLeftOffset = new float2(queueCol * (IconWidth + IconSpacing), 0);
+
+				var iconTopLeft = RenderOrigin + topLeftOffset;
+				var centerPosition = iconTopLeft;
+
+				WidgetUtils.DrawSHPCentered(icon.Image, centerPosition + 0.5f * iconSize, worldRenderer.Palette(actorData.BuildableInfo.IconPalette), 0.5f);
+
+				armyUnits.Add(actorInfo);
+				var rectangle = new Rectangle((int)iconTopLeft.X, (int)iconTopLeft.Y, (int)iconSize.X, (int)iconSize.Y);
+				armyIconsBounds.Add(rectangle);
+
+				var topColor = Color.FromArgb(0, 0, 0, 0);
+				var bottomColor = Color.FromArgb(228, 0, 0, 0);
+
+				// WidgetUtils.FillRectWithColor(new Rectangle((int)iconTopLeft.X, (int)(iconTopLeft.Y + iconSize.Y / 2), (int)iconSize.X, (int)iconSize.Y / 2), topColor, topColor, bottomColor, bottomColor);
```

Comment:
Keeping it here during the reviewing phase, uncomment to add the fading on the bottom of the icons to hide the text.